go-controller/pkg/ovn/base_network_controller_pods.go

```diff
@@ -331,31 +333,35 @@ func (bnc *BaseNetworkController) waitForNodeLogicalSwitchInCache(switchName str
 func (bnc *BaseNetworkController) addRoutesGatewayIP(pod *kapi.Pod, network *nadapi.NetworkSelectionElement,
 	podAnnotation *util.PodAnnotation, nodeSubnets []*net.IPNet) error {
 	if bnc.IsSecondary() {
-		topoType := bnc.TopologyType()
-		if topoType != ovntypes.Layer3Topology {
-			return fmt.Errorf("topology type %s not supported", topoType)
-		}
-		// secondary layer3 network, see if its network-attachment's annotation has default-route key.
+		// for secondary network, see if its network-attachment's annotation has default-route key.
 		// If present, then we need to add default route for it
 		podAnnotation.Gateways = append(podAnnotation.Gateways, network.GatewayRequest...)
-		for _, podIfAddr := range podAnnotation.IPs {
-			isIPv6 := utilnet.IsIPv6CIDR(podIfAddr)
-			nodeSubnet, err := util.MatchIPNetFamily(isIPv6, nodeSubnets)
-			if err != nil {
-				return err
-			}
-			gatewayIPnet := util.GetNodeGatewayIfAddr(nodeSubnet)
-			layer3NetConfInfo := bnc.NetConfInfo.(*util.Layer3NetConfInfo)
-			for _, clusterSubnet := range layer3NetConfInfo.ClusterSubnets {
-				if isIPv6 == utilnet.IsIPv6CIDR(clusterSubnet.CIDR) {
-					podAnnotation.Routes = append(podAnnotation.Routes, util.PodRoute{
-						Dest:    clusterSubnet.CIDR,
-						NextHop: gatewayIPnet.IP,
-					})
+		topoType := bnc.TopologyType()
+		switch topoType {
+		case ovntypes.Layer2Topology:
+			// no route needed for directly connected subnets
+			return nil
+		case ovntypes.Layer3Topology:
+			for _, podIfAddr := range podAnnotation.IPs {
+				isIPv6 := utilnet.IsIPv6CIDR(podIfAddr)
+				nodeSubnet, err := util.MatchIPNetFamily(isIPv6, nodeSubnets)
+				if err != nil {
+					return err
+				}
+				gatewayIPnet := util.GetNodeGatewayIfAddr(nodeSubnet)
+				layer3NetConfInfo := bnc.NetConfInfo.(*util.Layer3NetConfInfo)
+				for _, clusterSubnet := range layer3NetConfInfo.ClusterSubnets {
```

Comment:
but this may not be common among all topology type, and we can only assume this is specific to layer3 topology type. I don't see why we have to hide the cast.